### PR TITLE
jobs: add new columns to system.jobs table

### PR DESCRIPTION
**Pull request by dt**
This adds new columns to the jobs table for the properties of a jobs that need to be shown to a user in SHOW JOBS, so that it can avoid reading these from the binary payload, which is supposed to only be used by the job itself and not by humans or the jobs system.

Release note: none.
Epic: none.

### CODE REVIEW

**Review by RaduBerinde — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137601)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt and @msbutler)

___
*[`pkg/upgrade/upgrades/v25_1_add_jobs_tables.go` line 62 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137601#-OEKO_049fgC2HHWqEi3:-OEKO_049fgC2HHWqEi4:bqmx24h) ([raw file](https://github.com/cockroachdb/cockroach/blob/93ac7062a13ed61b7a2661fe27b2e0168095b602/pkg/upgrade/upgrades/v25_1_add_jobs_tables.go#L62)):*
> ```Go
> 
> // addJobsColumns adds new columns to system.jobs.
> func addJobsColumns(
> ```

[nit] These upgrades usually have an accompanying test. If you add one, please include a `clusterversion.SkipWhenMinSupportedVersionIsAtLeast(t, 25, 1)` and add the same to `TestPreparedTransactionsTable`

**Review by dt — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137601)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @msbutler and @RaduBerinde)

___
*[`pkg/upgrade/upgrades/v25_1_add_jobs_tables.go` line 62 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137601#-OEKO_049fgC2HHWqEi3:-OEKPjjr35T_GA6tJpQz:bd9s67s) ([raw file](https://github.com/cockroachdb/cockroach/blob/93ac7062a13ed61b7a2661fe27b2e0168095b602/pkg/upgrade/upgrades/v25_1_add_jobs_tables.go#L62)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] These upgrades usually have an accompanying test. If you add one, please include a `clusterversion.SkipWhenMinSupportedVersionIsAtLeast(t, 25, 1)` and add the same to `TestPreparedTransactionsTable`
</blockquote></details>

Since I’ve added gated usage of the state created by the migration in the same commit I think its behavior is well exercised by existing tests; do you think it needs an addition test here? I figured that would reduce to a test that “add column” adds a column which seems maybe uninteresting (and just adds to the version-specific churn to clean out later)

**Review by RaduBerinde — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137601)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt and @msbutler)

___
*[`pkg/upgrade/upgrades/v25_1_add_jobs_tables.go` line 62 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137601#-OEKO_049fgC2HHWqEi3:-OEKVvU--ush-R_F8aHw:bkmxptz) ([raw file](https://github.com/cockroachdb/cockroach/blob/93ac7062a13ed61b7a2661fe27b2e0168095b602/pkg/upgrade/upgrades/v25_1_add_jobs_tables.go#L62)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

Since I’ve added gated usage of the state created by the migration in the same commit I think its behavior is well exercised by existing tests; do you think it needs an addition test here? I figured that would reduce to a test that “add column” adds a column which seems maybe uninteresting (and just adds to the version-specific churn to clean out later)
</blockquote></details>

But the tests that exercise the gated usage do not start with an older cluster then upgrade it, no? They are essentially verifying the changes in `systemschema`, not this upgrade.

It seems like a very simple test but I remember one occasion where the addition of it caught a bug. In any case, as with any `[nit]` - your PR, your call :)

(also, ignore the comment about the existing `TestPreparedTransactionsTable`, I'm fixing it in #137617)

**Review by dt — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137601)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @msbutler and @RaduBerinde)

___
*[`pkg/upgrade/upgrades/v25_1_add_jobs_tables.go` line 62 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137601#-OEKO_049fgC2HHWqEi3:-OEKXiSsA9zK84--6JIm:b-ypf71b) ([raw file](https://github.com/cockroachdb/cockroach/blob/93ac7062a13ed61b7a2661fe27b2e0168095b602/pkg/upgrade/upgrades/v25_1_add_jobs_tables.go#L62)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

But the tests that exercise the gated usage do not start with an older cluster then upgrade it, no? They are essentially verifying the changes in `systemschema`, not this upgrade.

It seems like a very simple test but I remember one occasion where the addition of it caught a bug. In any case, as with any `[nit]` - your PR, your call :)

(also, ignore the comment about the existing `TestPreparedTransactionsTable`, I'm fixing it in #137617)
</blockquote></details>

There are numerous tests that run jobs in mixed version states, including `TestStartableJobMixedVersion` and actually many of the upgrade tests themselves since upgrades are run as jobs too (I also know this is exercised since I dropped one of the commas when reordering the ADD clauses and plenty of tests failed).

**Review by dt — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137601)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @msbutler and @RaduBerinde)

___
*[`pkg/upgrade/upgrades/v25_1_add_jobs_tables.go` line 62 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137601#-OEKO_049fgC2HHWqEi3:-OEKYRES-kSYfrCREUZu:b-ahmtao) ([raw file](https://github.com/cockroachdb/cockroach/blob/93ac7062a13ed61b7a2661fe27b2e0168095b602/pkg/upgrade/upgrades/v25_1_add_jobs_tables.go#L62)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

There are numerous tests that run jobs in mixed version states, including `TestStartableJobMixedVersion` and actually many of the upgrade tests themselves since upgrades are run as jobs too (I also know this is exercised since I dropped one of the commas when reordering the ADD clauses and plenty of tests failed).
</blockquote></details>

mixed version states _that bootstrap at min_ and then upgrade, specifically.

**Comment by RaduBerinde** [comment]
*[`pkg/upgrade/upgrades/v25_1_add_jobs_tables.go` line 62 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/137601#-OEKO_049fgC2HHWqEi3:-OEKZ1_yAzzLDorwHIwg:bn1wz0t) ([raw file](https://github.com/cockroachdb/cockroach/blob/93ac7062a13ed61b7a2661fe27b2e0168095b602/pkg/upgrade/upgrades/v25_1_add_jobs_tables.go#L62)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

mixed version states _that bootstrap at min_ and then upgrade, specifically.
</blockquote></details>

:thumbsup:

**Comment by msbutler** [comment]
it may be worth setting up the show jobs patch and running the mixed version backup test to check this migration is bug free before merging.

**Comment by dt** [comment]
> setting up the show jobs patch

I was planning to do the backfill patch next and then adding reads in SHOW comes last, since needs to gate on the backfill being done. I see your point about the proof being in the pudding in so far as until we are reading we don't know that the writes are correct, but where I hesitate is saying that we should make that the bar to merge the first pass, since that would mean coupling the whole thing into a monster change. So long as the first pass at writes doesn't break anything, I'm not _that_ worried about potential needing to come back and change a line a couple days later when we enable the reads, and prefer that to keeping a giant stack of diffs on a branch that rebases poorly due versions being a common sequence across all ongoing work.

File: pkg/jobs/update.go
Comment (dt) [reply pkg/jobs/update.go:342]: `FinalResumeError` is machine-readable rather than human readable so it can stay in a proto in `job_info` i.e. where it is now; it isn't and won't be used by `SHOW`

**Comment by dt** [comment]
>  drop the second descriptive string column entirely from the jobs system

went ahead and did this, and feel happier about it already.

**Review by msbutler — APPROVED** [review]
+1 on removing statement. Hope the test failures are easy to address.

**Comment by dt** [comment]
TFTR!

bors r+

**Pull request merged**
